src/visualization/classification/extra_trees_vis.py

```diff
@@ -0,0 +1,147 @@
+from PyQt5.QtWidgets import QWidget, QHBoxLayout, QVBoxLayout, QPushButton, QLabel, QGroupBox
+import numpy as np
+import pandas as pd
+import pygraphviz as pgv
+from typing import List, Dict, Optional
+
+import QGraphViz
+from QGraphViz.Engines import Dot
+from QGraphViz.QGraphViz import QGraphViz
+from QGraphViz.DotParser import Graph, GraphType
+
+
+class ExtraTreesStepsVisualization(QWidget):
+    def __init__(self, data: pd.DataFrame, algorithms_steps: List[str], is_animation: bool):
+        super().__init__()
+
+        self.data = data
+        self.steps = algorithms_steps
+        self.is_animation = is_animation
+
+        self.layout = QVBoxLayout(self)
+        self.graphs = [self.make_graph(dot_string) for dot_string in self.steps]
+        self.current_graph = 1
+
+        # graph section
+        self.graph_group = QGroupBox()
+        self.graph_group.setTitle("Graph")
+        self.graph_group_layout = QVBoxLayout(self.graph_group)
+        self.graph_group_layout.addWidget(self.graphs[self.current_graph - 1])
+        self.layout.addWidget(self.graph_group, 1)
+
+        # control panel
+        self.control_panel_layout = QHBoxLayout()
+        self.left_button = QPushButton("PREV")
+        self.left_button.clicked.connect(partial(self.click_listener, 'prev'))
+        self.right_button = QPushButton("NEXT")
+        self.right_button.clicked.connect(partial(self.click_listener, 'next'))
+        self.num_label = QLabel(f"Actual graph: {self.current_graph}")
+        self.description = QLabel("Red -> FALSE\nGreen -> TRUE")
+        self.random_button = QPushButton("Random graph")
+        self.random_button.clicked.connect(partial(self.click_listener, 'random'))
+        self.control_panel_layout.addWidget(self.description)
+        self.control_panel_layout.addStretch()
+        self.control_panel_layout.addWidget(self.num_label)
+        self.control_panel_layout.addStretch()
+        self.control_panel_layout.addWidget(self.left_button)
+        self.control_panel_layout.addWidget(self.random_button)
+        self.control_panel_layout.addWidget(self.right_button)
+        self.layout.addLayout(self.control_panel_layout, 0)
+
+    def postprocess_label(self, label: Optional[str]):
+        if label is None:
+            return ''
+        label = label[1:-1]
+        label = label.replace('&gt;', '>')
+        if '<br/>' not in label:
+            return label
+        return '\n'.join(label.split('<br/>'))
+        
+    def dict_of_param(self, data: Optional[str]) -> Dict:
+        if data is None:
+            return {}
+        res = {}
+        params = data.split(', ') if ', ' in data else data
+        for param in params:
+            key, value = param.split('=', 1)
+            if key == 'fillcolor':
+                value = value[1:-1]
+            if key == 'label':
+                value = self.postprocess_label(value)
+            res[key] = value
+        return res
+
+    def make_graph(self, data: str):
```

Comment:
I've been trying to comprehend this method for quite a while and I give up for today haha 

why is `data` a giant string that we have to parse here? I thought it was saved in this format for the library to draw it automatically, it doesn't?

Reply:
Yes, this string is parsed by pygraphviz to use graphviz engine to get position of nodes. To render the graph I write parser because there are some bugs in QGraphViz dot parser.